src/coreclr/jit/lowerxarch.cpp

```diff
@@ -847,384 +847,472 @@ void Lowering::LowerPutArgStk(GenTreePutArgStk* putArgStk)
 {
     assert(tree->OperIs(GT_CAST));
 
-    GenTree*  castOp     = tree->AsCast()->CastOp();
-    var_types castToType = tree->CastToType();
-    var_types dstType    = castToType;
-    var_types srcType    = castOp->TypeGet();
+    GenTree*  castOp  = tree->AsCast()->CastOp();
+    var_types dstType = tree->CastToType();
+    var_types srcType = castOp->TypeGet();
 
     // force the srcType to unsigned if GT_UNSIGNED flag is set
     if (tree->IsUnsigned())
     {
         srcType = varTypeToUnsigned(srcType);
     }
 
-    // We should not see the following casts unless directly supported by hardware,
-    // as they are expected to be lowered appropriately or converted into helper calls by front-end.
-    //   srcType = float/double                    castToType = * and overflow detecting cast
-    //       Reason: must be converted to a helper call
-    //   srcType = float/double,                   castToType = ulong
-    //       Reason: must be converted to a helper call
-    //   srcType = float/double,                   castToType = byte/sbyte/ushort/short
-    //       Reason: must have intermediate cast to int
-    //   srcType = uint                            castToType = float/double
-    //       Reason: uint -> float/double = uint -> long -> float/double
     if (varTypeIsFloating(srcType))
     {
+        // Overflow casts should have been converted to helper call in morph.
         noway_assert(!tree->gtOverflow());
+        // Small types should have had an intermediate int cast inserted in morph.
         assert(!varTypeIsSmall(dstType));
-        assert(castToType != TYP_ULONG || comp->canUseEvexEncodingDebugOnly());
+        // Long types should have been handled by helper call or in DecomposeLongs on x86.
+        assert(!varTypeIsLong(dstType) || TargetArchitecture::Is64Bit);
     }
     else if (srcType == TYP_UINT)
     {
-        assert(castToType != TYP_FLOAT || comp->canUseEvexEncodingDebugOnly());
+        // uint->float casts should have an intermediate cast to long unless
+        // we have the EVEX unsigned conversion instructions available.
+        assert(dstType != TYP_FLOAT || comp->canUseEvexEncodingDebugOnly());
     }
 
-#if defined(TARGET_AMD64)
-    // Handle saturation logic for X64
-    // Let InstructionSet_AVX10v2 pass through since it can handle the saturation
-    if (varTypeIsFloating(srcType) && varTypeIsIntegral(dstType) && !varTypeIsSmall(dstType) &&
+#ifdef FEATURE_HW_INTRINSICS
+    if (varTypeIsFloating(srcType) && varTypeIsIntegral(dstType) &&
         !comp->compOpportunisticallyDependsOn(InstructionSet_AVX10v2))
     {
-        // We should have filtered out float -> long conversion and
-        // converted it to float -> double -> long conversion.
-        assert((dstType != TYP_LONG) || (srcType != TYP_FLOAT));
+        // If we don't have AVX10v2 saturating conversion instructions for
+        // floating->integral, we have to handle the saturation logic here.
```

Comment:
This implementation is a complete rewrite, so it's best read top to bottom rather than compared against the current code.